cmd/proxygenerator/interceptor.go

```diff
@@ -109,39 +109,65 @@ func NewPayloadVisitorInterceptor(options PayloadVisitorInterceptorOptions) (grp
 
 		err := invoker(ctx, method, req, response, cc, opts...)
 		if err != nil {
-			if options.Inbound != nil {
-				stat, ok := status.FromError(err)
-				if ok {
-					// user provided payloads can sometimes end up in the status details of
-					// gRPC errors, make sure to visit those as well
-					newStatus := status.New(stat.Code(), stat.Message())
-					var newDetails []protoadapt.MessageV1
-					for _, detail := range stat.Details() {
-						detailAny, ok := detail.(*anypb.Any)
-						if ok && ({{ range $i, $name := .GrpcPayload }}{{ if $i }} || {{ end }}detailAny.MessageName() == "{{$name}}"{{ end }}) {
-							err = VisitPayloads(ctx, detailAny, *options.Inbound)
-							if err != nil {
-								return err
-							}
+			return parseGrpcPayload(ctx, err, options.Inbound)
+		}
+
+		if resMsg, ok := response.(proto.Message); ok && options.Inbound != nil {
+			return VisitPayloads(ctx, resMsg, *options.Inbound)	
+		}
+		
+		return nil
+	}, nil
+}
+
+func parseGrpcPayload(ctx context.Context, err error, inbound *VisitPayloadsOptions) error {
+	if inbound != nil {
+		stat, ok := status.FromError(err)
+		if ok {
+			// user provided payloads can sometimes end up in the status details of
+			// gRPC errors, make sure to visit those as well
+			newStatus := status.New(stat.Code(), stat.Message())
+			var newDetails []protoadapt.MessageV1
+			for _, detail := range stat.Details() {
+				detailAny, okAny := detail.(*anypb.Any)
+				detailV1, okDetail := detail.(protoadapt.MessageV1)
+				payloadTypes := []string{ {{ range $i, $name := .GrpcPayload }}{{ if $i }}, {{ end }}"{{$name}}"{{ end }} }
+				if okAny {
+					if slices.Contains(payloadTypes, string(detailAny.MessageName())) {
+						err = VisitPayloads(ctx, detailAny, *inbound)
+						if err != nil {
+							return err
 						}
-						newDetails = append(newDetails, detailAny)
 					}
-					newStatus, err = newStatus.WithDetails(newDetails...)
+					//newDetails = append(newDetails, detailAny)
+					newStatus, err = newStatus.WithDetails(detailAny)
 					if err != nil {
 						return err
 					}
-					return newStatus.Err()
+				} else if okDetail {
+					msg := protoadapt.MessageV2Of(detailV1)
+					if slices.Contains(payloadTypes, string(msg.ProtoReflect().Descriptor().FullName())) {
+						err = VisitPayloads(ctx, msg, *inbound)
+					}
+					newDetails = append(newDetails, detailV1)
+				} else {
+					// stat.Details() returned an error
+					detailErr, ok := detail.(error)
+					if !ok {
+						return fmt.Errorf("unreachable")
+					}
+						return detailErr
```

Comment:
Can probably remove this, I'm ok if we just don't append details that don't match message v1 or anypb.Any (they will always be anypb.Any anyways looking at code).

Actually, looking back at this, if I were you, I'd operate on the raw status proto. So my function may look like:

```go
func parseGrpcPayload(ctx context.Context, err error, inbound *VisitPayloadsOptions) error {
  if err == nil || inbound == nil {
    return err
  }
  s, ok := status.FromError(err)
  if !ok {
    return err
  }
  p := s.Proto()
  for _, detail := range p.Details {
    if slices.Contains(payloadTypes, string(detailAny.MessageName())) {
      if vErr := VisitPayloads(ctx, detail, *inbound); vErr != nil {
        return vErr
      }
    }
  }
  return status.ErrorProto(p)
}
```